File: 11_动态规划/背包问题/技巧/撤销背包/完全背包/UnboundedKnapsackRemovable.ts
/**
 * 可撤销完全背包，用于求解方案数/可行性.
 */
class UnboundedKnapsackRemovable {
  private readonly _dp: number[]
  private readonly _maxWeight: number
  private readonly _mod?: number | undefined

  constructor(maxWeight: number, mod?: number, dp?: number[]) {
    if (dp != undefined) {
      this._dp = dp
    } else {
      this._dp = Array(maxWeight + 1).fill(0)
      this._dp[0] = 1
    }
    this._maxWeight = maxWeight
    this._mod = mod
  }

  /**
   * 添加一个重量为weight的物品.
   * @complexity O(maxWeight)
   */
  add(weight: number): void {
    if (this._mod == undefined) {
      for (let i = weight; i <= this._maxWeight; i++) {
        this._dp[i] += this._dp[i - weight]
      }
    } else {
      for (let i = weight; i <= this._maxWeight; i++) {
        this._dp[i] = (this._dp[i] + this._dp[i - weight]) % this._mod
      }
    }
  }

  /**
   * 移除一个重量为weight的物品.需要保证weight物品存在.
   * @complexity O(maxWeight)
   */
  remove(weight: number): void {
    if (this._mod == undefined) {
      for (let i = this._maxWeight; i >= weight; i--) {
        this._dp[i] -= this._dp[i - weight]
      }
    } else {
      for (let i = this._maxWeight; i >= weight; i--) {
        this._dp[i] = (this._dp[i] - this._dp[i - weight]) % this._mod
      }
    }
  }

  /**
   * 查询组成重量为weight的物品有多少种方案.
   * @complexity O(1)
   */
  query(weight: number): number {
    if (weight < 0 || weight > this._maxWeight) return 0
    if (this._mod == undefined) return this._dp[weight]
    if (this._dp[weight] < 0) this._dp[weight] += this._mod
    return this._dp[weight]
  }

  copy(): UnboundedKnapsackRemovable {
    return new UnboundedKnapsackRemovable(this._maxWeight, this._mod, this._dp.slice())
  }
}

export { UnboundedKnapsackRemovable }

if (require.main === module) {
  const knapsack = new UnboundedKnapsackRemovable(10)
  knapsack.add(1)
  knapsack.add(2)
  knapsack.remove(2)
  console.log(knapsack.query(10))
}

File: 11_动态规划/背包问题/技巧/撤销背包/完全背包/UnboundedKnapsackRemovable.go
package main

// 可撤销完全背包，用于求解方案数/可行性.
type UnboundedKnapsackRemovable struct {
	dp        []int
	maxWeight int
	mod       int
}

// maxWeight: 背包最大容量.
// mod: 模数，传入-1表示不需要取模.
func NewUnboundedKnapsackRemovable(maxWeight int, mod int) *UnboundedKnapsackRemovable {
	dp := make([]int, maxWeight+1)
	dp[0] = 1
	return &UnboundedKnapsackRemovable{
		dp:        dp,
		maxWeight: maxWeight,
		mod:       mod,
	}
}

// 添加一个重量为weight的物品.
func (ks *UnboundedKnapsackRemovable) Add(weight int) {
	if ks.mod == -1 {
		for i := weight; i <= ks.maxWeight; i++ {
			ks.dp[i] += ks.dp[i-weight]
		}
	} else {
		for i := weight; i <= ks.maxWeight; i++ {
			ks.dp[i] = (ks.dp[i] + ks.dp[i-weight]) % ks.mod
		}
	}
}

// 移除一个重量为weight的物品.需要保证weight物品存在.
func (ks *UnboundedKnapsackRemovable) Remove(weight int) {
	if ks.mod == -1 {
		for i := ks.maxWeight; i >= weight; i-- {
			ks.dp[i] -= ks.dp[i-weight]
		}
	} else {
		for i := ks.maxWeight; i >= weight; i-- {
			ks.dp[i] = (ks.dp[i] - ks.dp[i-weight]) % ks.mod
		}
	}
}

// 查询组成重量为weight的物品有多少种方案.
// !注意需要特判重量为0.
func (ks *UnboundedKnapsackRemovable) Query(weight int) int {
	if weight < 0 || weight > ks.maxWeight {
		return 0
	}
	if ks.mod == -1 {
		return ks.dp[weight]
	}
	if ks.dp[weight] < 0 {
		ks.dp[weight] += ks.mod
	}
	return ks.dp[weight]
}

func (ks *UnboundedKnapsackRemovable) Copy() *UnboundedKnapsackRemovable {
	dp := append(ks.dp[:0:0], ks.dp...)
	return &UnboundedKnapsackRemovable{
		dp:        dp,
		maxWeight: ks.maxWeight,
		mod:       ks.mod,
	}
}
